### PR TITLE
(feat) - O3-3096 service queues - onfiguration-driven queue table

### PR DESCRIPTION
**Pull request by chibongho**
## Requirements

- [ ] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [ ] My work includes tests or is validated by existing tests.

## Summary

This is a draft proposal of how to address the problem describe in [O3-3096](https://openmrs.atlassian.net/browse/O3-3096):

We need to configure the queue table in different contexts. For example, there can be a view of queue entries from all queues, a view of queue entries in a specific queue, or even a view of queue entries in a specific queue AND status. Depending on the context, we want to configure the queue table to display different columns. Additionally, the queue table usually provides a way to perform actions on each queue entry. The set of available actions could be specific to a queue, specific to a status, or specific to a queue+status combination.

The PR proposes a solution that allows us to specify `TableDefinition`s in the config. Each `TableDefinition` specifies a list of columns to use based on an `appliedTo` criteria, which can be applied to specific queue, a specific status, a specific queue+status combination, or (in the case when `appliedTo` not specified) to any table. 

We predefine a set of `QueueTableColumn`s that we should reasonably support (like name, priority, status) and how they should get rendered. Some of these columns can further be configured. (For example, this PR shows how we can configure the `patientId` column to display a particular `patientIdentifierType`.) We will also define another `QueueTableColumn` (not currently in this PR) that only renders an extension slot, allowing implementors to custom-define whatever column they want.

Also not currently in this PR yet, but we will also need more flexibility in how we define actions we can perform to each queue entry based on its current queue and status. We can predefined a set of `QueueEntryAction`s that we should reasonably support (like editing / transitioning / ending / undoing queue entries). Some of these actions can further be customized (For example, we can define a quick action that tranistions a queue entry to a particular status). In addition, we can also define another `QueueEntryAction` that only renders an extension slot, allowing implementors to custom-define whatever action they want. We can use a very similar approach that we used for configuring columns and apply it to `QueueEntryAction`s
 

## Screenshots

Table listing queue entries from multiple queues:
image

Table listing queue entries from a specific queue. Note that the columns rendered are different.
image

## Other

[O3-3096]: https://openmrs.atlassian.net/browse/O3-3096?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/esm-service-queues-app/src/queue-table/default-queue-table.component.tsx
Comment (chibongho) [inline packages/esm-service-queues-app/src/queue-table/default-queue-table.component.tsx:71]: This (and similar columns arrays) are no longer hard-coded.

File: packages/esm-service-queues-app/src/queue-table/cells/columns.resource.ts
Comment (chibongho) [inline packages/esm-service-queues-app/src/queue-table/cells/columns.resource.ts:48]: Big switch statement here to mapping columnType (defined in config) to actual columns that are used to render header and cells of a column in the DataTable. The switch statement enumerates all the predefined columns we support. Implementators wishing to have a custom column can use the `extension-column` columnType.

File: packages/esm-service-queues-app/src/queue-table/cells/columns.resource.ts
Comment (chibongho) [inline packages/esm-service-queues-app/src/queue-table/cells/columns.resource.ts:18]: Function to determine the columns to display for a queue table of a particular queue + status

File: packages/esm-service-queues-app/src/config-schema.ts
Comment (chibongho) [inline packages/esm-service-queues-app/src/config-schema.ts:7]: Definition of default columns to show in the queue table of the service queues app main page.

File: packages/esm-service-queues-app/src/config-schema.ts
Comment (chibongho) [inline packages/esm-service-queues-app/src/config-schema.ts:50]: 2 different `TableDefinition`s here. The first one is for illustration purpose, to show that we can define the columns for a specific queue. The second one applies to all tables.

File: packages/esm-service-queues-app/src/queue-table/cells/columns.resource.ts
Comment (mseaton) [inline packages/esm-service-queues-app/src/queue-table/cells/columns.resource.ts:30]: Do we need to get the `t` function here and pass it down, rather than just get it in the getColumnFromDefinition method directly?

File: packages/esm-service-queues-app/src/queue-table/cells/columns.resource.ts
Comment (mseaton) [inline packages/esm-service-queues-app/src/queue-table/cells/columns.resource.ts:53]: The custom `header` here should _also_ be run through the translation function, no?  Potentially also with another optional property on `columnDef` to indicate which module the header translation string is defined in?

File: packages/esm-service-queues-app/src/queue-table/cells/queue-table-patient-id-cell.component.tsx
Comment (mseaton) [inline packages/esm-service-queues-app/src/queue-table/cells/queue-table-patient-id-cell.component.tsx:4]: `Id` -> `Identifier` search/replace in this file

File: packages/esm-service-queues-app/src/types/index.ts
Comment (mseaton) [inline packages/esm-service-queues-app/src/types/index.ts:433]: This supports a message code right?  Might want to indicate this can be a translation string or actual column value

File: packages/esm-service-queues-app/translations/en.json
Comment (mseaton) [inline packages/esm-service-queues-app/translations/en.json:172]: We probably don't actually want these message strings to be removed, and there may be some we need to see added from the different built-in columns.  There is a trick here that you need to do to ensure that the build scripts behave correctly - you can see how this is done in the root component of patient-chart-app.  Look in the codebase for a comment that reads something like:  `DO NOT REMOVE THIS COMMENT` and `THE TRANSLATION KEYS AND VALUES USED IN THE COMMON LIB IS WRITTEN HERE`

**Review by mseaton — COMMENTED** [review]
@chibongho looks great.  A few comments.  I see this is still in Draft - would be good to get a sense of what is remaining before moving it into a state that could be merged.

File: packages/esm-service-queues-app/src/config-schema.ts
Comment (chibongho) [reply packages/esm-service-queues-app/src/config-schema.ts:251]: This is the same name as config used in the appointments app and patient-chart. This is technically a new schema, but I think I should retain the same name for consistency.

**Comment by chibongho** [comment]
> @chibongho looks great. A few comments. I see this is still in Draft - would be good to get a sense of what is remaining before moving it into a state that could be merged.

@mseaton There are a few more `QueueTableColumn`s currently marked as TODO that I'd like to define as part of this PR. They are relatively easy to do. The configuration of actions will be trickier and probably warrant a separate PR.

File: packages/esm-service-queues-app/translations/en.json
Comment (chibongho) [reply packages/esm-service-queues-app/translations/en.json]: I got around this by changing the `extract-translations` git hook to read in *.ts files.

**Comment by chibongho** [comment]
This is now ready for review. Thanks.

File: packages/esm-service-queues-app/src/queue-table/cells/queue-table-patient-age-cell.component.tsx
Comment (mseaton) [inline packages/esm-service-queues-app/src/queue-table/cells/queue-table-patient-age-cell.component.tsx:5]: This should be renamed to `QueueTablePatientAgeCell`

File: packages/esm-service-queues-app/src/queue-table/cells/queue-table-patient-age-cell.component.tsx
Comment (mseaton) [inline packages/esm-service-queues-app/src/queue-table/cells/queue-table-patient-age-cell.component.tsx:16]: Same, rename once above is renamed

File: packages/esm-service-queues-app/src/queue-table/cells/queue-table-patient-identifier-cell.component.tsx
Comment (mseaton) [inline packages/esm-service-queues-app/src/queue-table/cells/queue-table-patient-identifier-cell.component.tsx:12]: `getPatientId` -> `getPatientIdentifier`

File: packages/esm-service-queues-app/src/queue-table/cells/queue-table-patient-identifier-cell.component.tsx
Comment (mseaton) [inline packages/esm-service-queues-app/src/queue-table/cells/queue-table-patient-identifier-cell.component.tsx:14]: We likely want to check beyond this.  If a patient has > 1 identifier of the same type, then we should prioritize a) any that are "preferred = true" first, then likely the most recently created second.  Ideally the API could give you this, but otherwise should be done here.

File: packages/esm-service-queues-app/src/queue-table/cells/queue-table-patient-age-cell.component.tsx
Comment (mseaton) [inline packages/esm-service-queues-app/src/queue-table/cells/queue-table-patient-age-cell.component.tsx:8]: I have seen logic around O3 that attempts to standardize some of this in places.  Not sure if we want to try to use that here, or if that should be a different column type.  The logic I've seen is something like "if the patient is < 1 month old, show age in days, if the patient is < 2 years old show age in months, else show age in years".  Don't rely on that exactly, but it's something like that.  @ibacher / @denniskigen do you remember?

File: packages/esm-service-queues-app/src/queue-table/cells/queue-table-patient-identifier-cell.component.tsx
Comment (mseaton) [inline packages/esm-service-queues-app/src/queue-table/cells/queue-table-patient-identifier-cell.component.tsx:21]: `QueueTablePatientIdCell` -> `QueueTablePatientIdentifierCell`

File: packages/esm-service-queues-app/src/queue-table/cells/queue-table-queue-cell.component.tsx
Comment (mseaton) [inline packages/esm-service-queues-app/src/queue-table/cells/queue-table-queue-cell.component.tsx:9]: This is really `queueName` right?  Let's rename the component to this

File: packages/esm-service-queues-app/src/queue-table/cells/queue-table-visit-start-time-cell.component.tsx
Comment (mseaton) [inline packages/esm-service-queues-app/src/queue-table/cells/queue-table-visit-start-time-cell.component.tsx:6]: visit is theoretically nullable on queueEntry, so may want to make this a `?.`

File: packages/esm-service-queues-app/src/queue-table/cells/queue-table-visit-start-time-cell.component.tsx
Comment (mseaton) [inline packages/esm-service-queues-app/src/queue-table/cells/queue-table-visit-start-time-cell.component.tsx:9]: Copy/pasted code needs to be renamed.

File: packages/esm-service-queues-app/translations/en.json
Comment (mseaton) [inline packages/esm-service-queues-app/translations/en.json:199]: Why is this getting removed?

**Review by mseaton — APPROVED** [review]
Looks great to me @chibongho .  I'm approving, but there are some minor changes to be made, mostly where you've copied/pasted code around and failed to rename everything.  I don't need to re-review assuming you clean up the comments indicated.

I would be good to get at least one or two reviews from some of the more experienced react developers

File: packages/esm-service-queues-app/src/queue-table/cells/columns.resource.ts
Comment (ibacher) [reply packages/esm-service-queues-app/src/queue-table/cells/columns.resource.ts]: There's also the framework's 	`translateFrom()` function, which might be easier than maintaining two `t` functions (since presumably the defaults should still come from this namespace).

File: packages/esm-service-queues-app/translations/en.json
Comment (chibongho) [reply packages/esm-service-queues-app/translations/en.json:199]: Turns out I forgot to have a case statement for this column. Thanks for catching.

**Pull request merged**
